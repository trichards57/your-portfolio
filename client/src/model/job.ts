export type Outcome =
  | "DischargedOnScene"
  | "NotFound"
  | "StoodDown"
  | "Conveyed"
  | "Other";

export interface NewJob {
  age?: number;
  blueLights: boolean;
  category: 1 | 2 | 3 | 4 | 5;
  drove: boolean;
  gender?: "Male" | "Female";
  notes: string;
  outcome: Outcome;
  reflectionFlag: boolean;
  shift: string;
}

export interface JobSummary {
  age?: number;
  category: 1 | 2 | 3 | 4 | 5;
  gender?: "Male" | "Female";
  reflectionFlag: boolean;
}
