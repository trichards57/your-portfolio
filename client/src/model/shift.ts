export type RoleType = "EAC" | "AFA" | "CRU";

export interface Shift {
  crewMate?: string;
  date: string;
  duration: number;
  event: string;
  id: string;
  location?: string;
  role: RoleType;
}

export interface ShiftSummary extends Shift {
  loggedCalls: number;
}

export type NewShift = Omit<Shift, "id">;
