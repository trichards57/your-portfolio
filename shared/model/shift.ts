export interface ShiftSummary {
  id: number;
  date: string;
  event: string;
  location: string;
  duration: number;
  role: "EAC" | "AFA" | "CRU";
  crewMate: string;
  loggedCalls: number;
}
