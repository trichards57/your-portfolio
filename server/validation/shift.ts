import { isValid, parseISO } from "date-fns";
import { ValidationError } from "./validation-error";
import { NewShift } from "../../shared/model/shift";

export function validateNewShift(
  item: Partial<NewShift>
): [ValidationError<NewShift>[], NewShift | undefined] {
  let res: ValidationError<NewShift>[] = [];

  if (!item.date || !isValid(parseISO(item.date)))
    res.push({
      error: "invalid date",
      fieldName: "date",
    });

  if (!item.event || item.event.trim() === "")
    res.push({
      error: "missing event name",
      fieldName: "event",
    });

  if (!(item.role == "AFA" || item.role == "CRU" || item.role == "EAC"))
    res.push({
      error: "invalid role",
      fieldName: "role",
    });

  if (typeof item.duration != "number" || !Number.isFinite(item.duration))
    res.push({
      error: "invalid duration",
      fieldName: "duration",
    });

  return [res, item as NewShift];
}
