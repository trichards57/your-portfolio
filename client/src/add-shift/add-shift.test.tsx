import { fireEvent, render, waitFor } from "@testing-library/react";
import React from "react";
import ReactDOM from "react-dom";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { useAuth0 } from "@auth0/auth0-react";
import { useHistory } from "react-router-dom";
import { NewShift } from "../model/shift";
import { AddShiftBase } from ".";

jest.mock("@auth0/auth0-react");
jest.mock("react-router-dom", () => ({
  useHistory: jest.fn(),
}));
const testToken = "abcdefg";

const endPoint = "/api/LogShift";

const testUpdate: NewShift = {
  date: "2021-01-02",
  duration: 6.5,
  event: "SWASFT Shift",
  role: "CRU",
  crewMate: "New Crew Mate",
  location: "New Location",
};

const server = setupServer(
  rest.post(endPoint, (req, res, ctx) => {
    if (req.headers.get("authorization") === `Bearer ${testToken}`) {
      const update = JSON.parse(req.body?.toString() ?? "") as NewShift;

      expect(update).toEqual(testUpdate);

      return res(ctx.status(200));
    }
    return res(ctx.status(401));
  })
);

beforeAll(() => server.listen());

beforeEach(() =>
  (useAuth0 as jest.Mock).mockReturnValue({
    getAccessTokenSilently: jest
      .fn()
      .mockReturnValue(Promise.resolve(testToken)),
  })
);

afterEach(() => server.resetHandlers());

afterAll(() => server.close());

it("renders without crashing", () => {
  const div = document.createElement("div");

  ReactDOM.render(<AddShiftBase />, div);

  ReactDOM.unmountComponentAtNode(div);
});

it("renders correctly", async () => {
  const res = render(<AddShiftBase />);

  expect(res.asFragment()).toMatchSnapshot();
});

it("displays an invalid date properly", async () => {
  const res = render(<AddShiftBase />);

  const box = res.getByLabelText("Date", {
    exact: false,
  });

  fireEvent.change(box, {
    target: { value: "" },
  });

  const alertId = box.getAttribute("aria-describedby");
  expect(box).toBeInvalid();

  expect(res.container.querySelector(`#${alertId}`)).toHaveTextContent(
    "You need to enter a date"
  );
  expect(res.getByRole("button", { name: "Save" })).toBeDisabled();
});

it("displays an invalid duration properly", async () => {
  const res = render(<AddShiftBase />);

  const box = res.getByRole("spinbutton", {
    name: "Hours",
  });

  fireEvent.change(box, {
    target: { value: "" },
  });

  const alertId = box.getAttribute("aria-describedby");
  expect(box).toBeInvalid();

  expect(res.container.querySelector(`#${alertId}`)).toHaveTextContent(
    "You need to enter a shift length"
  );
  expect(res.getByRole("button", { name: "Save" })).toBeDisabled();
});

it("displays an invalid event name properly", async () => {
  const res = render(<AddShiftBase />);

  const box = res.getByRole("textbox", {
    name: "Event",
  });

  fireEvent.change(box, {
    target: { value: "a" },
  });

  fireEvent.change(box, {
    target: { value: "" },
  });

  const alertId = box.getAttribute("aria-describedby");
  expect(box).toBeInvalid();

  expect(res.container.querySelector(`#${alertId}`)).toHaveTextContent(
    "You need to name this shift"
  );
  expect(res.getByRole("button", { name: "Save" })).toBeDisabled();
});

it("submits data to the server when save is clicked", async () => {
  const testPush = jest.fn();

  (useHistory as jest.Mock).mockReturnValue({
    push: testPush,
  });

  const res = render(<AddShiftBase />);

  fireEvent.change(res.getByLabelText("Date", { exact: false }), {
    target: { value: testUpdate.date },
  });
  fireEvent.change(res.getByRole("spinbutton", { name: "Hours" }), {
    target: { value: testUpdate.duration.toString() },
  });
  fireEvent.change(res.getByRole("textbox", { name: "Event" }), {
    target: { value: testUpdate.event },
  });
  fireEvent.change(res.getByRole("textbox", { name: "Location" }), {
    target: { value: testUpdate.location },
  });
  fireEvent.change(res.getByLabelText("Role", { exact: false }), {
    target: { value: testUpdate.role },
  });
  fireEvent.change(res.getByRole("textbox", { name: "Crew Mate" }), {
    target: { value: testUpdate.crewMate },
  });

  const saveButton = res.getByRole("button", { name: "Save" });

  fireEvent.click(saveButton);

  expect(saveButton).toBeDisabled();
  await waitFor(() => expect(testPush).toBeCalledWith("/home"));
  expect(saveButton).toBeEnabled();
});

it("displays alert when save fails", async () => {
  const testPush = jest.fn();

  (useHistory as jest.Mock).mockReturnValue({
    push: testPush,
  });

  const res = render(<AddShiftBase />);

  server.use(rest.post(endPoint, (_, rs, ctx) => rs(ctx.status(500))));

  fireEvent.change(res.getByLabelText("Date", { exact: false }), {
    target: { value: testUpdate.date },
  });
  fireEvent.change(res.getByRole("spinbutton", { name: "Hours" }), {
    target: { value: testUpdate.duration.toString() },
  });
  fireEvent.change(res.getByRole("textbox", { name: "Event" }), {
    target: { value: testUpdate.event },
  });
  fireEvent.change(res.getByRole("textbox", { name: "Location" }), {
    target: { value: testUpdate.location },
  });
  fireEvent.change(res.getByLabelText("Role", { exact: false }), {
    target: { value: testUpdate.role },
  });
  fireEvent.change(res.getByRole("textbox", { name: "Crew Mate" }), {
    target: { value: testUpdate.crewMate },
  });

  const saveButton = res.getByRole("button", { name: "Save" });

  fireEvent.click(saveButton);

  expect(saveButton).toBeDisabled();
  await waitFor(() => res.getByRole("alert"));

  const alert = res.getByRole("alert");
  expect(alert).toHaveTextContent("problem speaking to the server");

  expect(testPush).not.toBeCalled();

  expect(saveButton).toBeEnabled();
});
