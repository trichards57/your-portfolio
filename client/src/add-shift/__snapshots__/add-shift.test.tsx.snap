// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-18"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-20 mui-fixed MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          aria-label="open drawer"
          class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-21 MuiIconButton-colorInherit MuiIconButton-edgeStart"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap"
        >
          Your Portfolio
        </h6>
      </div>
    </header>
    <nav
      aria-label="mailbox folders"
      class="makeStyles-drawer-19"
    />
    <main
      class="makeStyles-content-24"
    >
      <div
        class="makeStyles-toolbar-22"
      />
      <div
        class="MuiPaper-root makeStyles-root-15 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <h2
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        >
          Add Shift
        </h2>
        <form
          novalidate=""
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-8 MuiGrid-grid-md-3 MuiGrid-grid-lg-2"
            >
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-item-26"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled Mui-required Mui-required"
                  data-shrink="true"
                  for="date"
                  id="date-label"
                >
                  Date
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiFilledInput-input"
                    id="date"
                    required=""
                    type="date"
                    value="2021-02-26"
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-2"
            >
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-item-26"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled Mui-required Mui-required"
                  data-shrink="true"
                  for="hours"
                  id="hours-label"
                >
                  Hours
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiFilledInput-input"
                    id="hours"
                    required=""
                    type="number"
                    value="8"
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-7 MuiGrid-grid-lg-8"
            >
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-item-26"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled Mui-required Mui-required"
                  data-shrink="false"
                  for="event"
                  id="event-label"
                >
                  Event
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiFilledInput-input"
                    id="event"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-item-26"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
                  data-shrink="false"
                  for="location"
                  id="location-label"
                >
                  Location
                </label>
                <div
                  class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiFilledInput-input"
                    id="location"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 MuiGrid-grid-lg-2"
            >
              <div
                class="MuiFormControl-root makeStyles-item-26"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled Mui-required Mui-required"
                  data-shrink="true"
                  for="role"
                  id="role-label"
                >
                  Role
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                >
                  <select
                    aria-invalid="false"
                    class="MuiSelect-root MuiSelect-select MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
                    id="role"
                    required=""
                  >
                    <option
                      value="EAC"
                    >
                      EAC
                    </option>
                    <option
                      value="AFA"
                    >
                      AFA
                    </option>
                    <option
                      value="CRU"
                    >
                      CRU
                    </option>
                  </select>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-8 MuiGrid-grid-lg-10"
            >
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-item-26"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
                  data-shrink="false"
                  for="crew-mate"
                  id="crew-mate-label"
                >
                  Crew Mate
                </label>
                <div
                  class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiFilledInput-input"
                    id="crew-mate"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-flex-end MuiGrid-grid-xs-12"
            >
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
                  disabled=""
                  id="save-button"
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    <span
                      class="MuiButton-startIcon MuiButton-iconSizeMedium"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
                        />
                      </svg>
                    </span>
                    Save
                  </span>
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
